09/tree23/node.go

```diff
@@ -0,0 +1,475 @@
+	} else if ((n.length == 2) && (x < n.keys[1])) || (n.length == 1) {
+		return n.sons[1].Search(x)
+	} else if n.length == 2 {
+		return n.sons[2].Search(x)
+	}
+
+	return nil
+}
+
+func (n *node[T]) Remove(x T) *node[T] {
+	item := n.Search(x)
+
+	if item == nil {
+		return n
+	}
+
+	var minNode *node[T]
+	if item.keys[0] == x {
+		minNode = item.sons[1].searchMin()
+	} else {
+		minNode = item.sons[2].searchMin()
+	}
+
+	if minNode != nil {
+		var z T
+		if x == item.keys[0] {
+			z = item.keys[0]
+		} else {
+			z = item.keys[1]
+		}
+		z, minNode.keys[0] = minNode.keys[0], z
+		item = minNode
+	}
+
+	item.removeFromNode(x)
+	return item.fix() // Вызываем функцию для восстановления свойств дерева.
+}
+
+func (n *node[T]) isLeaf() bool {
+	return n.sons[0] == nil && n.sons[1] == nil && n.sons[2] == nil && n.sons[3] == nil
+}
+
+func (n *node[T]) insertToNode(x T) {
+	n.keys[n.length] = x
+	n.length++
+	n.sort()
+}
+
+func (n *node[T]) split() *node[T] {
+	if n.length < 3 {
+		return n
+	}
+
+	x := &node[T]{
+		parent: n.parent,
+		sons:   [4]*node[T]{n.sons[0], n.sons[1]},
+		keys:   [3]T{n.keys[0]},
+		length: 1,
+	}
+	y := &node[T]{
+		parent: n.parent,
+		sons:   [4]*node[T]{n.sons[2], n.sons[3]},
+		keys:   [3]T{n.keys[2]},
+		length: 1,
+	}
+	if x.sons[0] != nil {
+		x.sons[0].parent = x
+	}
+	if x.sons[1] != nil {
+		x.sons[1].parent = x
+	}
+	if y.sons[0] != nil {
+		y.sons[0].parent = y
+	}
+	if y.sons[1] != nil {
+		y.sons[1].parent = y
+	}
+
+	if n.parent != nil {
+		n.parent.insertToNode(n.keys[1])
+
+		if n.parent.sons[0] == n {
+			n.parent.sons[0] = nil
+		} else if n.parent.sons[1] == n {
+			n.parent.sons[1] = nil
+		} else if n.parent.sons[2] == n {
+			n.parent.sons[2] = nil
+		}
+
+		if n.parent.sons[0] == nil {
+			n.parent.sons[3] = n.parent.sons[2]
+			n.parent.sons[2] = n.parent.sons[1]
+			n.parent.sons[1] = y
+			n.parent.sons[0] = x
+		} else if n.parent.sons[1] == nil {
+			n.parent.sons[3] = n.parent.sons[2]
+			n.parent.sons[2] = y
+			n.parent.sons[1] = x
+		} else {
+			n.parent.sons[3] = y
+			n.parent.sons[2] = x
+		}
+
+		return n.parent
+	} else {
+		x.parent = n
+		y.parent = n
+
+		n.sons = [4]*node[T]{x, y, nil, nil}
+		n.keys[0] = n.keys[1]
+		n.length = 1
+
+		return n
+	}
+}
+
+func (n *node[T]) sort() {
+	switch n.length {
+	case 1:
+		return
+	case 2:
+		if n.keys[0] > n.keys[1] {
+			n.keys[0], n.keys[1] = n.keys[1], n.keys[0]
+		}
+	case 3:
+		if n.keys[0] > n.keys[1] {
+			n.keys[0], n.keys[1] = n.keys[1], n.keys[0]
+		}
+		if n.keys[0] > n.keys[2] {
+			n.keys[0], n.keys[2] = n.keys[2], n.keys[0]
+		}
+		if n.keys[1] > n.keys[2] {
+			n.keys[1], n.keys[2] = n.keys[2], n.keys[1]
+		}
+	}
+}
+
+func (n *node[T]) findKey(x T) bool {
+	for i := range n.length {
+		if n.keys[i] == x {
+			return true
+		}
+	}
+	return false
+}
+
+func (n *node[T]) searchMin() *node[T] {
+	if n == nil {
+		return n
+	}
+	if !(n.sons[0] == nil) {
+		return n.searchMin()
+	} else {
+		return n.sons[0].searchMin()
+	}
+}
```

Comment:
первый баг тут:
непонятно, зачем зацикленно вызывать поиск минимума для той же самой ноды, если есть потомки, и если есть, логично было бы вызвать поиск минимума именно в потомке
```suggestion
func (n *node[T]) searchMin() *node[T] {
	if n == nil {
		return n
	}
	if n.sons[0] == nil {
		return n
	} 
	
	return n.sons[0].searchMin()
}
```